### PR TITLE
[POAE7-2813] ArrowSourceNode Refactor to Register Output Batch

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Move output batch registration to the beginning of query_func.

### Why are the changes needed?

To support vectorize translators, moving output batch registration to the beginning will reduce copy and move between intermediate array and output array. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs

### Which label does this PR belong to?

INFRA

### CODE REVIEW

File: src/cider/exec/nextgen/context/CodegenContext.h [hunk 1]
```diff
@@ -225,6 +224,14 @@ class CodegenContext {
   // returns an index used for retrieving the charset at runtime
   int registerTrimStringOperCharMap(const std::string& trim_chars);
 
+ private:
```
Comment (winningsix) [inline src/cider/exec/nextgen/context/CodegenContext.h:227]: why needs to move here out of previous private section?

**Comment by winningsix** [comment]
> vectorize

Do we have some design doc regards to op vectorization?

File: src/cider/exec/nextgen/context/CodegenContext.h (hunk 1 above)
Comment (bigPYJ1151) [reply src/cider/exec/nextgen/context/CodegenContext.h:227]: Just to make code in good order.

**Comment by bigPYJ1151** [comment]
> > vectorize
> 
> Do we have some design doc regards to op vectorization?

Not yet. Currently I will select some vectorizable expression and move them out of ProjectNode. Then generates their code as multiple tight loops with vectorized null processing.

File: cpp/src/cider/exec/nextgen/parsers/RelAlgEUParser.cpp [hunk 2]
```diff
@@ -145,9 +147,18 @@ OpPipeline toOpPipeline(RelAlgExecutionUnit& eu) {
   OpPipeline ops;
 
   InputAnalyzer analyzer(eu);
-  analyzer.run();
+  auto&& input_exprs = analyzer.run();
+
+  // Relpace ColumnVar in target_exprs with OutputColumnVar to distinguish input cols and
+  // output cols.
+  for (auto& expr : eu.shared_target_exprs) {
+    if (auto column_var_ptr = std::dynamic_pointer_cast<Analyzer::ColumnVar>(expr)) {
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/nextgen/parsers/RelAlgEUParser.cpp:155]: the top ColumnVar as OutputColumnVar?

File: cpp/src/cider/type/plan/ColumnExpr.h [hunk 3]
```diff
@@ -113,6 +113,19 @@ class ColumnVar : public Expr {
   ColumnInfoPtr col_info_;
 };
 
+class OutputColumnVar : public Expr {
+ public:
+  explicit OutputColumnVar(const std::shared_ptr<ColumnVar>& col)
+      : Expr(col->get_type_info()), col_(col) {}
+
+  JITExprValue& codegen(CodegenContext& context) override;
```
Comment (Ma-Jian1) [inline cpp/src/cider/type/plan/ColumnExpr.h:121]: not implemented?

File: cpp/src/cider/exec/nextgen/context/RuntimeContext.h [hunk 4]
```diff
@@ -57,8 +57,10 @@ class RuntimeContext {
     if (batch_holder_.empty()) {
       return nullptr;
     }
-    auto batch = batch_holder_.back().second.get();
+    auto batch = batch_holder_.front().second.get();
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply cpp/src/cider/exec/nextgen/context/RuntimeContext.h:60]: Right.

File: cpp/src/cider/type/plan/ColumnExpr.h (hunk 3 above)
Comment (bigPYJ1151) [reply cpp/src/cider/type/plan/ColumnExpr.h:121]: In .cpp file.

**Review by Ma-Jian1 — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Review by winningsix — APPROVED** [review]

**Pull request merged**